cmd/virt-chroot/tap-device-maker.go

```diff
@@ -2,50 +2,121 @@ package main
 
 import (
 	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
 	"strconv"
 	"strings"
+	"syscall"
 	"time"
 
 	"github.com/spf13/cobra"
 	"github.com/vishvananda/netlink"
 	"golang.org/x/sys/unix"
 )
 
-func createTapDevice(name string, owner uint, group uint, queueNumber int, mtu int) error {
-	tapDevice := &netlink.Tuntap{
-		LinkAttrs:  netlink.LinkAttrs{Name: name},
-		Mode:       unix.IFF_TAP,
-		NonPersist: false,
-		Queues:     queueNumber,
-		Owner:      uint32(owner),
-		Group:      uint32(group),
+func createTapDevice(name string, parentName string, owner uint, group uint, queueNumber int, mtu int) error {
+	if parentName == "" {
+		// Parent not specified. Create a tap interface.
+		tapDevice := &netlink.Tuntap{
+			LinkAttrs:  netlink.LinkAttrs{Name: name},
+			Mode:       unix.IFF_TAP,
+			NonPersist: false,
+			Queues:     queueNumber,
+			Owner:      uint32(owner),
+			Group:      uint32(group),
+		}
+
+		// when netlink receives a request for a tap device with 1 queue, it uses
+		// the MULTI_QUEUE flag, which differs from libvirt; as such, we need to
+		// manually request the single queue flags, enabling libvirt to consume
+		// the tap device.
+		// See https://github.com/vishvananda/netlink/issues/574
+		if queueNumber == 1 {
+			tapDevice.Flags = netlink.TUNTAP_DEFAULTS
+		}
+
+		// Device creation is retried due to https://bugzilla.redhat.com/1933627
+		// which has been observed on multiple occasions on CI runs.
+		const retryAttempts = 5
+		attempt, err := retry(retryAttempts, func() error {
+			return netlink.LinkAdd(tapDevice)
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create tap device named %s. Reason: %v", name, err)
+		}
+
+		if err := netlink.LinkSetMTU(tapDevice, mtu); err != nil {
+			return fmt.Errorf("failed to set MTU on tap device named %s. Reason: %v", name, err)
+		}
+
+		fmt.Printf("Successfully created tap device %s, attempt %d\n", name, attempt)
+		return nil
+	}
+
+	// Parent specified. Create a macvtap.
+	p, err := netlink.LinkByName(parentName)
+	if err != nil {
+		return fmt.Errorf("failed to lookup lowerDevice %q: %v", parentName, err)
+	}
+
+	macvtapDevice := &netlink.Macvtap{
+		Macvlan: netlink.Macvlan{
+			LinkAttrs: netlink.LinkAttrs{
+				Name:        name,
+				ParentIndex: p.Attrs().Index,
+				// we had crashes if we did not set txqlen to some value
+				TxQLen: p.Attrs().TxQLen,
+			},
+			Mode: netlink.MACVLAN_MODE_BRIDGE,
```

Comment:
Do you know if `VEGA` mode will work here ?

Reply:
Do you mean `VEPA`?
It will work only for connectivity to upper interface, but not to other macvlan interface used for DHCP:

![изображение](https://user-images.githubusercontent.com/7556217/234609224-d883d372-9e2a-4fb3-80b0-7d9a21d8a0a0.png)